### PR TITLE
Add generated assets to main compiler

### PR DESCRIPTION
**Pull request by lalexdotcom**
The assets generated by childCompilation (basically the JS files) are now added to the main compilation assets, allowing other Webpack plugins to be aware of it.

FIXES: #58

### CODE REVIEW

**Review by simoneb — COMMENTED** [review]
1. link the issue to the PR
2. write a test :)

**Comment by ShogunPanda** [comment]
Other than what Simone said, rest LGTM! Nice job!

**Comment by lalexdotcom** [comment]
OK.

1. How can I link the PR to the issue after creation?
2. Should I create new test or editing the main test by checking the assets is enough?

**Comment by ShogunPanda** [comment]
1. Just write a `Fixes: #NUMBER` in the PR description so that GitHub will link it.
2. Up to you, whatever you find most readable and comfortable.

**Comment by lalexdotcom** [comment]
Actually, the used hook is wrong, sometimes never get called...

**Comment by mcollina** [comment]
@ShogunPanda what test can we add for this feature?

**Comment by ShogunPanda** [comment]
Well, since the original problem was that the pino assets disappeared, maybe we can just check they don't when used in conjunction of the web pack cleaning plugin.

**Comment by lalexdotcom** [comment]
The test would be to check if the assets property of returned `stats` contains pino generated files....

I had to leave the computer, but I'll work on that by the end of the day... (my day, in Europe 😉)

**Comment by lalexdotcom** [comment]
Clean plugin was just the plug-in who led to find this issue, but other plugins may need to have accurate assets from compilation....

**Comment by lalexdotcom** [comment]
I'm actually stuck. When I add an asset to the main compilation assets, two files are generated with different hashes (one before adding the footer code, and one after)...

I have to spend more time on it, it was a bit too soon to send the PR. 😅

**Comment by lalexdotcom** [comment]
I finally had time to spend on this...
It's nearly a complete refactoring to be sure to use the correct hooks, but the documentation is so poor about child compilers and hooks... 😕

@gdorsi : as you wrote the PR to handle cache in this plugin, could you have a look at my code? I really don't understand how to handle the cache... and if you could explain to me how it works, I would really appreciate! 😇

**Comment by simoneb** [comment]
@lalexdotcom meanwhile it might be worth looking into why CI fails

**Comment by lalexdotcom** [comment]
You're right, it's linting issues: the variables used by cache handling were still there, and sometimes I prefer having no brackets for a simple `if`
I mus admit I didn't laucnh the formatting & linting tasks. It's done in my new commit...

File: src/index.js
Comment (mcollina) [inline src/index.js:5]: why is this commented out?

Comment (lalexdotcom) [reply src/index.js:5]: As I said in my commit, I couldn't find where to handle the cache in the refactoring: https://github.com/pinojs/pino-webpack-plugin/pull/59#issuecomment-1386734928
To be able to emit assets from the child compilation, I had to use different webpack hooks...

Comment (lalexdotcom) [reply src/index.js:5]: Actually, it's nearly a complete refactoring, I thought it would be more readable this way:
- I got into the (too light) Webpack hooks documentation, and used more appropriate hooks to handle additional assets
- I thought that having functions declared further in code but used only once doesn't help readability. I used the function body directly in the webpack hooks

Please let me know if you think I should have kept the previous code "style"...

**Review by gdorsi — COMMENTED** [review]
Well done @lalexdotcom, switching to `webpack.EntryPlugin` is a great improvement!

WIth `webpack.EntryPlugin` the new files are no more considered external so [the cache is already handled by webpack](https://github.com/webpack/webpack/blob/4b4ca3bb53f36a5b8fc6bc1bd976ed7af161bd80/lib/Compilation.js#L2387) (TBH I don't understand 100% the referenced code, but since the tests are passing I think there is no need to go deeper)

As general advice for your future contributions, try to keep the style changes in a separate PR because they make the review process way harder.

File: src/index.js
Comment (lalexdotcom) [reply src/index.js:89]: Hey @gdorsi

I thought that avoiding writing intermediate assets to the file system and keeping them in memory would be a good idea...

Actually, not all the running tests are passing: **the cached build test doesn't pass** (this is why I tagged you in my comment). The second build is generating new files...

Noted for changing style in PRs. Actually, it helped me to understand the process of the plugin, so for me it was kinda part of the feature...

Could you take a look on the cache handling part? And if you could explain to me how it works...

Comment (lalexdotcom) [reply src/index.js:89]: And the EntryPlugin was already there as I remember (`addExternalFile` and `trackInclusions` functions)

Comment (lalexdotcom) [reply src/index.js:89]: Still not for me:
```
test/cached-build.test.js
  it should correctly generate all required pino files when the cache is enabled on Webpack
    ✓ should not error
    ✓ should not error
    ✓ expect falsey value
    ✓ expect truthy value
    1) expect truthy value
    ✓ expect truthy value
    ✓ expect truthy value
    ✓ should match pattern provided
    ✓ should match pattern provided
```

Maybe there's a cache folder I did'nt clear, but I don't see any...

Comment (gdorsi) [reply src/index.js:89]: > Noted for changing style in PRs. Actually, it helped me to understand the process of the plugin, so for me it was kinda part of the feature...

I also do that when I'm facing new code :smile:, but once you have reached a working solution it is good to go back and minimize the number of changes.

It's just a matter to help each other to get the best possible result, especially for this kind of library that is both complex and rarely updated.

Comment (lalexdotcom) [reply src/index.js:89]: This is the file structure when I output the cahed-build test to not temp folder
image

Comment (lalexdotcom) [reply src/index.js:89]: > > Noted for changing style in PRs. Actually, it helped me to understand the process of the plugin, so for me it was kinda part of the feature...
> 
> I also do that when I'm facing new code 😄, but once you have reached a working solution it is good to go back and minimize the number of changes.
> 
> It's just a matter to help each other to get the best possible result, especially for this kind of library that is both complex and rarely updated.

Yes, I'll know that for my next PRs...
it's my first contribution to a project, thanks for understanding

Comment (gdorsi) [reply src/index.js:89]: The tests on CI [are passing](https://github.com/pinojs/pino-webpack-plugin/actions/runs/3948483588/jobs/6758620789#step:6:20) so it should be something on your local setup.

Comment (gdorsi) [reply src/index.js:89]: > I thought that avoiding writing intermediate assets to the file system and keeping them in memory would be a good idea...

This kind of optimization may come already from webpack, in order to avoid issues on edge cases (e.g. the dev server) I suggest reverting to the standard one.

Comment (lalexdotcom) [reply src/index.js:89]: That's weird: I tried to run it in a devcontainer to be independant from local system, and the test still doesn't pass...

Comment (lalexdotcom) [reply src/index.js:89]: > > I thought that avoiding writing intermediate assets to the file system and keeping them in memory would be a good idea...
> 
> This kind of optimization may come already from webpack, in order to avoid issues on edge cases (e.g. the dev server) I suggest reverting to the standard one.

Actually, we have two files generated if you don't use this: the original one is written to the file system (without the compatibility footer), but not deleted when you create the main compilation asset with the footer added... we would have to handle manually the original file deletion....

Comment (lalexdotcom) [reply src/index.js:89]: Actually, it could be a good thing to rewrite the tests not to only check that the expected files are here, but also that ONLY the expected files are here... also by checking that each file is an asset on the resulting webpack stats, etc... 

(and more human friendly description would be nice too 😇)

I'll could make a PR if you think it's a good idea...

Comment (gdorsi) [reply src/index.js:89]: > Could you take a look on the cache handling part? And if you could explain to me how it works.

When the cache is enabled `trackInclusions` isn't called for the files that are not changed.
So the `handleCache` method kicks in to restore the missing modules from the cache.

> That's weird: I tried to run it in a devcontainer to be independant from local system, and the test still doesn't pass...

So it may be related to the environment or the dependencies.

I'm struggling to understand what you have done, could you either explain your solution or remove the style changes?

Comment (gdorsi) [reply src/index.js:89]: Found the issue, the tests are passing because of the `memoryFs` that is created at the module level.

**Review by gdorsi — CHANGES_REQUESTED** [review]
There are too many unrelated changes here and this PR is too hard to be reviewed properly.

Let's focus on what we need to fix the issue and keep the rest for future contributions.

File: src/index.js
Comment (lalexdotcom) [reply src/index.js:89]: The main problem in the current code is:
- The child compiler is fully independant: it adds the compatibility footer after emitting its assets to files, then store the generated paths.
- The main compiler add content (banner) to the entrypoints files, but is actually not aware of what files have been generated by the child compiler (except in the banner added)
- Some code could be in a more appropriate hook (`processAssets` with `PROCESS_ASSETS_STAGE_ADDITIONS` stage to add content to an asset for example)

**This leads to the issue #58** : other plugins never know that this plugin created new files as they are not parts of the "assets" property of the compilation stats. And for '"destructive" plugins (like CleanWebpackPlugin), this means the new files are deleted

To be able to avoid that, we have to use a different flow: we cannot add the assets on the child compiler callback as those assets are already emmited, that's why I have to do this in the child compiler's `chunkAsset` and `processAssets`hook.
I can do this again by letting the functions which are outside of the main plugin, but I'd have to change their bodies and arguments, it seem to me that it's gonna be as hard to review it...

As you like, I don't wan't to interfere with the project...

Comment (gdorsi) [reply src/index.js:89]: Thanks for the explanation :smile: 

Switching back to `childCompiler.outputFileSystem = compiler.outputFileSystem` breaks a few tests on my local machine.

To be sure that this is a real issue, could you try the same on your local?

Comment (lalexdotcom) [reply src/index.js:89]: I confirm that it breaks some tests: as there is now several file like "pino-file-xxxx.js", and the test is choosing the first one in alphabetical order...

**Comment by ShogunPanda** [comment]
The code refactoring looks good from what I can see. The refactoring was huge and thus I trust the CI.
I think we can resolve conflicts in #58 in a separate PR. Once tests passes everywhere, let's land this and move on.

@lalexdotcom Thanks for the PR, this was huge.

**Comment by ShogunPanda** [comment]
@gdorsi @lalexdotcom Since CI is already passing, I wait for the green light from you to land this.

**Comment by lalexdotcom** [comment]
> The code refactoring looks good from what I can see. The refactoring was huge and thus I trust the CI. I think we can resolve conflicts in #58 in a separate PR. Once tests passes everywhere, let's land this and move on.
> 
> @lalexdotcom Thanks for the PR, this was huge.

Hi @ShogunPanda, thanks for your feedback.
@gdorsi pointed out that the CI seem to pass because of an issue with memoryFs https://github.com/pinojs/pino-webpack-plugin/pull/59#discussion_r1073520408

This PR already solves #58, but in my local config AND in a local devcontainer (Node 18-bullseye for VSCode), cache tests don't pass... this is a regression and maybe it would not be wasted time for reliability to spend more time on this, maybe on different user machine, using isolated environment?

For semver, can adding a new dependency be considered as a breaking change? If not, no breaking change...

**Comment by simoneb** [comment]
> For semver, can adding a new dependency be considered as a breaking change? If not, no breaking change...

thanks for the clarification on this point.

File: src/index.js
Comment (gdorsi) [reply src/index.js:89]: Ah ok, I missed [this](https://github.com/pinojs/pino-webpack-plugin/pull/59#discussion_r1073482267) so now I understand why the memory fs is useful (would be worth a comment in the code).

Good call!

Have you find out why the tests are not passing on your local?

If not I could try to help.

Can you share your node version and the lockfile?

**Comment by gdorsi** [comment]
> @gdorsi @lalexdotcom Since CI is already passing, I wait for the green light from you to land this.

@ShogunPanda there is still the issue of a test failing on @lalexdotcom local.

Once we have sorted that out the PR is good to go for me.

File: src/index.js
Comment (lalexdotcom) [reply src/index.js:89]: The tests are now passing on my local computer **BUT SHOULDN'T**: two 'first-xxxx.js' and 'abc/cde/second-xxxxx.js' files are generated...
As `readdirSync` is returning filenames in alphabetical order, depending on the hash, tests could succeed when it shouldn't... for example they should fail if there's several file starting with 'first-'...

A way to help me would be to test with no temp dir, look at the output and see why the file generated on the second build with cache enabled has a different name: as I said, I don't really understand how the webpack cache works and there's poor documentation on that... does the childCompiler use cache?

Comment (lalexdotcom) [reply src/index.js:92]: > Ah ok, I missed [this](https://github.com/pinojs/pino-webpack-plugin/pull/59#discussion_r1073482267) so now I understand why the memory fs is useful (would be worth a comment in the code) and why the tests fail if it is disabled.

You're perfectly right, I added a comment on that in a new commit, let me know if it's clear enough

Comment (gdorsi) [reply src/index.js:92]: > look at the output and see why the file generated on the second build with cache enabled has a different name

If the name is generated using the content hash it is normal for them to have different names.

>  I don't really understand how the webpack cache works and there's poor documentation on that... does the childCompiler use cache?

The poor docs is what make writing Webpack plugins so hard.
I did this one year ago and don't remember these parts, only that you need to check the Webpack source code to understand what's happening.

Comment (lalexdotcom) [reply src/index.js:92]: > > look at the output and see why the file generated on the second build with cache enabled has a different name
> 
> If the name is generated using the content hash it is normal for them to have different names.

What I don't understand is that file content hasn't changed, so the hash shouldn't have either...
I'm gonna try to dive deeper in the cache handling, but don't know when I'll be able to fix this...

**Comment by lalexdotcom** [comment]
I think I nailed it! 😎

When using cache, the `buildModule` hook isn't called if the module is cached. As I was using this hook to populate the childCompiler entries, the second build was generating another `__bundlerPathsOverrides` object (empty, because no module was rebuilt)... different object ➡️ different file content ➡️ different name hash ➡️ different file 😓

I'm now using the `finishModules` hook to populate the childCompiler entries in one loop on all the modules, even the one that were previously cached... it allows also to not have to handle cache by hand...

**Comment by lalexdotcom** [comment]
As this PR was originally made to fix #58, i also updated the clean-plugin.test to perform 2 builds and be sure that cold start and restart have the same behavior, which is required for watch mode...

**Comment by mcollina** [comment]
@ShogunPanda PTAL

**Review by ShogunPanda — APPROVED** [review]

**Comment by simoneb** [comment]
@ShogunPanda @lalexdotcom are we happy with this PR then? I assume this needs to go in before #83 ?

**Comment by lalexdotcom** [comment]
@simoneb OK for me.

I would say we can merge #59 then I'll make another commit to #83 with updated clean-plugin.test.js...

**Comment by mcollina** [comment]
go for it!

**Pull request merged**
